### PR TITLE
REP-207 Add filter config

### LINKED ISSUE #207: Create a new standalone Filter configuration which a Site can use to describe what data it wants
**Describe the task**
A new Solr core needs to be created that holds filter information for sites. This will be similar to the current replication-configuration core but will have only the following fields.

    id
    version
    related site id
    filter (cql string)
    name
    description
    suspended (boolean)

A manager to read/write this core also needs to be created.

### PR DESCRIPTION
**Pull request by kcover**
#### What does this PR do?
Adds a filter config object to replication along with the persistence manager and tests to go with it.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@peterhuffer @clockard @paouelle 

#### How should this be tested? (List steps with links to updated documentation)
CI Build 

#### What are the relevant tickets?
Fixes #207 

#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

### CODE REVIEW

File: replication-api/src/main/java/com/connexta/replication/api/data/Filter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.data;
+
+import javax.annotation.Nullable;
+
+/** A filter for retrieving data from a specific site. */
+public interface Filter extends Persistable {
```
Comment (clockard) [inline replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: Should filters also have a priority? When we are putting items in our priority queues I would imagine that the filter priority would be used as the item priority in the queue.

Comment (kcover) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: Is that going to be used atm? I think we'll want to do it eventually but I'm not sure if it should be done right now. I could just add a priority that should be a number 1-10 if we know that's how we want to do it.

Comment (clockard) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: I think it will be used when we switch over to queues in a couple sprints. We could wait till then but that's just one more thing to do and potentially a different version of the config.

Comment (clockard) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: The 1-10 matches up with the queue broker api priority ranges so I think that would be ok. The other option would be to use the enums defined in the api but I think i like just a straight 1-10 better.

Comment (paouelle) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: I would prefer we be consistent and reuse the `PriorityLevel` enum by moving it into a common package to be shared by both the API and the queue.

If not then we should remove that enum and opt for a 1-10 everywhere.

Either way I opt for consistency.

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+          && Objects.equals(name, pojo.name)
+          && Objects.equals(description, pojo.description);
+    }
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    return String.format(
+        "FilterPojo[id=%s, version=%s, siteId=%s, filter=%s, name=%s, description=%s]",
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java]: ✏️ missing `suspended`

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    return filter;
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public String getDescription() {
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]: ✏️ Should return `Optional<String>`

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    return description;
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]:
```suggestion
    return Optional.ofNullable(description);
```

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    return pojo.setSiteId(siteId)
+        .setFilter(filter)
+        .setName(name)
+        .setDescription(description)
+        .setSuspended(isSuspended);
+  }
+
+  @Override
+  protected void readFrom(FilterPojo pojo) {
+    super.readFrom(pojo);
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]: ❗️ Add the following after `super.readFrom()`
```
if (pojo.getVersion() < FilterPojo.MINIMUM_VERSION) {
      throw new ReplicationPersistenceException(
          "unsupported "
              + FilterImpl.TYPE
              + " version: "
              + pojo.getVersion()
              + " for object: "
              + getId());
    } // do support pojo.getVersion() > CURRENT_VERSION for forward compatibility
```

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    setOrFailIfNullOrEmpty("siteId", pojo::getSiteId, this::setSiteId);
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]: ✏️ I ended up moving the remainder of this method into a method named `readFromCurrentAndFutureVersion(FilterPojo pojo)` and called it from it in preparation of future versions support

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    return filterStreamOf(filterRepository.findAll().spliterator());
+  }
+
+  @Override
+  public void save(Filter filter) {
+    if (filter instanceof FilterImpl) {
+      filterRepository.save(((FilterImpl) filter).writeTo(new FilterPojo()));
+    } else {
+      throw new IllegalArgumentException(
+          "Expected a ReplicatorConfigImpl but got a " + filter.getClass().getSimpleName());
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterManagerImpl.java]: `FilterImpl`

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  @Test
+  public void hashcode() {
+    FilterPojo filterPojo2 = makeDefaultPojo();
+    assertThat(filterPojo.hashCode(), is(filterPojo2.hashCode()));
+    filterPojo2.setName("numbatwo");
+    assertThat(filterPojo.hashCode(), not(filterPojo2.hashCode()));
+  }
+
+  @Test
+  public void testToString() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java]: ✏️ Not sure we should test the `toString()`

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+  private static final String NAME = "name";
+
+  private static final String DESCRIPTION = "description";
+
+  private FilterPojo filterPojo;
+
+  @Before
+  public void setup() {
+    filterPojo = makeDefaultPojo();
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java]: ❓ Why do this in a setup method? Why not just initialize the attribute above? By design JUnit will create a new instance for each test it runs

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  @Test
+  public void save() {
+    FilterPojo filterPojo = makeDefaultPojo();
+    filterManager.save(new FilterImpl(filterPojo));
+    verify(filterRepository).save(eq(filterPojo));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void saveIllegalObject() {
+    filterManager.save(new TestFilter());
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: ❓ Can we not just use Mockito to create a mock of `Filter`?

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterImplTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    filter.setSuspended(true);
+    assertThat(filter.getSiteId(), is(SITE_ID));
+    assertThat(filter.getFilter(), is(FILTER));
+    assertThat(filter.getName(), is(NAME));
+    assertThat(filter.getDescription(), is(DESCRIPTION));
+    assertThat(filter.isSuspended(), is(true));
+  }
+
+  @Test
+  public void readFrom() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterImplTest.java]: ❗️ We should add test cases to verify the failures that can occur when we read and validate the data we receive

File: replication-api/src/main/java/com/connexta/replication/api/data/Filter.java (hunk 1 above)
Comment (clockard) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: Agree we should try to be consistent. I think the 1-10 gives us more flexibility but I am ok either way. I do think we should include the priority in this iteration though.

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterImplTest.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    filter = new FilterImpl(pojo);
+    assertThat(filter.getSiteId(), is(SITE_ID));
+    assertThat(filter.getFilter(), is(FILTER));
+    assertThat(filter.getName(), is(NAME));
+    assertThat(filter.getDescription(), is(DESCRIPTION));
+    assertThat(filter.isSuspended(), is(true));
+  }
+
+  @Test
+  public void writeTo() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterImplTest.java]: ❗️ We should add test cases to verify the failures that can occur when we write and validate the data

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/spring/FilterRepositoryIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+    repo.save(POJO2);
+    repo.save(POJO3);
+
+    assertThat(repo.count(), equalTo(3L));
+    repo.deleteAll();
+    assertThat(repo.count(), equalTo(0L));
+  }
+
+  @Test
+  public void testFindBySiteId() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/spring/FilterRepositoryIntegrationTest.java]: ❗️ Add another pojo with `SITE_ID` to verify you get more than one match back.

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java (hunk 10 above)
Comment (kcover) [reply replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: I need the instanceof statement to be false. I wasn't sure how to achieve that with a mock.

File: solr-schema/src/main/resources/solr/conf/replication-solr-managed-schema.xml [hunk 14]
```diff
@@ -165,17 +165,17 @@
...
-    <field name="suspended" type="boolean" indexed="true" stored="true" required="false"
-           multiValued="false" />
     <field name="last_metadata_modified" type="pdate" indexed="true" stored="false" required="false"
            multiValued="false" />
 
+    <!-- replication_filter fields -->
+    <field name="name" type="string" indexed="true" stored="true" required="false" multiValued="false" />
+    <field name="site_id" type="string" indexed="true" stored="true" required="false" multiValued="false" />
+    <field name="filter" type="string" indexed="true" stored="true" required="false" multiValued="false" />
+    <field name="suspended" type="boolean" indexed="true" stored="true" required="false" multiValued="false" />
```
[the remark this answers is not in the document]
Comment (kcover) [reply solr-schema/src/main/resources/solr/conf/replication-solr-managed-schema.xml]: Now that I think about it I don't think so, I'll change this and filter to not be indexed and make them unsearchable in the impls.

File: replication-api/src/main/java/com/connexta/replication/api/data/Filter.java (hunk 1 above)
Comment (paouelle) [reply replication-api/src/main/java/com/connexta/replication/api/data/Filter.java]: Let's go with a byte and fix it to 1-10. Let's make the change to the queue api to remove the enum and set it as a byte there as well. This should be a simple change that we could do at the same time for consistency.

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java (hunk 10 above)
Comment (paouelle) [reply replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: Mock the interface. This will dynamically create an object class that implements it but that will be different than the one internally.

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java (hunk 9 above)
Comment (kcover) [reply replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java]: I thought I would be changing the values of filterPojo in tests when I wrote this. Turns out I didn't have to. I'll change it

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java (hunk 2 above)
Comment (peterhuffer) [reply replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java]: ✏️ Can add a toString test and use the ToStringVerifier to catch missing fields like this.

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    return siteId;
+  }
+
+  /**
+   * Sets the site id for this filter
+   *
+   * @param siteId the id of the site to associate with this filter.
+   * @return this for chaining
+   */
+  public FilterPojo setSiteId(String siteId) {
```
[the remark this answers is not in the document]
Comment (kcover) [reply replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java]: Except for suspended and priority right?

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java (hunk 6 above)
Comment (kcover) [reply replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]: should that be `readFromCurrentAndFutureVersions` or `readFromCurrentOrFutureVersion`? I'm thinking the latter

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (paouelle) [reply replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java]: yeah. all the object-based ones, not the primitive :-)

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java (hunk 6 above)
Comment (paouelle) [reply replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]: or ... `readFromCurrentOrFutureVersionAsLongAsItIsGreaterThanTheMinimumSupportedVersion()`

😄 

I like the later. @peterhuffer Take note of the method name change

File: solr-schema/src/main/resources/solr/conf/replication-solr-managed-schema.xml [hunk 16]
```diff
@@ -173,8 +173,9 @@
     <!-- replication_filter fields -->
     <field name="name" type="string" indexed="true" stored="true" required="false" multiValued="false" />
     <field name="site_id" type="string" indexed="true" stored="true" required="false" multiValued="false" />
-    <field name="filter" type="string" indexed="true" stored="true" required="false" multiValued="false" />
-    <field name="suspended" type="boolean" indexed="true" stored="true" required="false" multiValued="false" />
+    <field name="filter" type="string" indexed="false" stored="true" required="false" multiValued="false" />
+    <field name="suspended" type="boolean" indexed="false" stored="true" required="false" multiValued="false" />
+    <field name="priority" type="pint" indexed="true" stored="true" required="false" multiValued="false" />
```
Comment (kcover) [inline solr-schema/src/main/resources/solr/conf/replication-solr-managed-schema.xml]: We should not be indexing this.

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    readFromCurrentOrFutureVersion(pojo);
+  }
+
+  private void readFromCurrentOrFutureVersion(FilterPojo pojo) {
+    setOrFailIfNullOrEmpty("siteId", pojo::getSiteId, this::setSiteId);
+    setOrFailIfNullOrEmpty("filter", pojo::getFilter, this::setFilter);
+    setOrFailIfNullOrEmpty("name", pojo::getName, this::setName);
+    this.description = pojo.getDescription();
+    this.isSuspended = pojo.isSuspended();
+    this.priority = readPriority(pojo);
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterImpl.java]:
```suggestion
    this.priority = Math.min(Math.max(pojo.getPriority(), 1), 10);
```

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterManagerImpl.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+import com.connexta.replication.api.impl.persistence.spring.FilterRepository;
+import com.connexta.replication.api.persistence.FilterManager;
+import java.util.Spliterator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/** Performs CRUD operations for {@link Filter}s. */
+public class FilterManagerImpl implements FilterManager {
+
+  FilterRepository filterRepository;
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterManagerImpl.java]: ✏️ private and final

File: replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+          && Objects.equals(name, pojo.name)
+          && Objects.equals(description, pojo.description);
+    }
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    return String.format(
+        "FilterPojo[id=%s, version=%s, siteId=%s, filter=%s, name=%s, description=%s, suspended=%s, priority=%s]",
```
Comment (paouelle) [inline replication-api-impl/src/main/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojo.java]: ✏️ `%d` for priority

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    assertThat(filter.getName(), is(NAME));
+    assertThat(filter.getPriority(), is(PRIORITY));
+  }
+
+  @Test(expected = NotFoundException.class)
+  public void getFilterNotFound() {
+    when(filterRepository.findById(anyString())).thenReturn(Optional.empty());
+    filterManager.get("id");
+  }
+
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: ❗️ We should add a test case when it fails because the persitence layer throws a persistence exception

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+  @Test
+  public void objects() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: ❗️ We should add a test case when it fails because the persitence layer throws a persistence exception on one of the object

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+    assertThat(filter.getId(), is(ID));
+    assertThat(filter.getSiteId(), is(SITE_ID));
+    assertThat(filter.getDescription().get(), is(DESCRIPTION));
+    assertThat(filter.getFilter(), is(FILTER));
+    assertThat(filter.getName(), is(NAME));
+    assertThat(filter.getPriority(), is(PRIORITY));
+  }
+
+  @Test
+  public void save() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: ❗️ We should add a test case when it fails because the persitence layer throws a persistence exception

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  @Test
+  public void remove() {
+    filterManager.remove("id");
+    verify(filterRepository).deleteById("id");
+  }
+
+  @Test
+  public void filtersForSite() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: ❗️ We should add a test case when it fails because the persitence layer throws a persistence exception on one of the object

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+  @Test
+  public void equalsWithDifferentPriority() {
+    FilterPojo filterPojo2 = makeDefaultPojo().setPriority((byte) 5);
+    assertFalse(filterPojo.equals(filterPojo2));
+    assertFalse(filterPojo2.equals(filterPojo));
+  }
+
+  @Test
+  public void equalsWithDifferentObject() {
```
Comment (paouelle) [inline replication-api-impl/src/test/java/com/connexta/replication/api/impl/persistence/pojo/FilterPojoTest.java]: we should add a test when called with null

File: replication-api/src/main/java/com/connexta/replication/queue/TaskInfo.java [hunk 25]
```diff
@@ -20,11 +20,11 @@
 /** Represents a task that can be queued for later processing by a worker. */
 public interface TaskInfo {
   /**
-   * Gets the priority for this task.
+   * Gets the priority for this task, represented as a number between 1 and 10.
    *
    * @return the priority for the task
    */
-  public PriorityLevel getPriority();
```
Comment (paouelle) [inline replication-api/src/main/java/com/connexta/replication/queue/TaskInfo.java]: ❗️ Remove the PriorityLevel enum

File: replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java (hunk 21 above)
Comment (kcover) [reply replication-api-impl/src/test/java/com/connexta/replication/api/impl/data/FilterManagerImplTest.java]: You mean the repository right? Why would we test the repository here?

File: replication-api/src/main/java/com/connexta/replication/queue/TaskInfo.java (hunk 25 above)
Comment (kcover) [reply replication-api/src/main/java/com/connexta/replication/queue/TaskInfo.java]: I did apparently git thinks I changed it to be the filterManager interface.
